File: MATTHEW/leetcode/AIcoding/5.Qwen-3/java_solutions/1914-cyclically-rotating-a-grid.java
public class Solution {

import java.util.*;

public class Solution {
    public List<List<Integer>> rotateGrid(int[][] grid, int k) {
        int rows = grid.length;
        int cols = grid[0].length;
        List<List<Integer>> result = new ArrayList<>();

        for (int i = 0; i < rows; i++) {
            result.add(new ArrayList<>());
        }

        int top = 0, bottom = rows - 1, left = 0, right = cols - 1;

        while (top <= bottom && left <= right) {
            List<Integer> cycle = new ArrayList<>();

            // Traverse from left to right on top row
            for (int i = left; i <= right; i++) {
                cycle.add(grid[top][i]);
            }

            // Traverse from top+1 to bottom on right column
            for (int i = top + 1; i <= bottom; i++) {
                cycle.add(grid[i][right]);
            }

            // Traverse from right-1 to left on bottom row, if top != bottom
            if (top < bottom) {
                for (int i = right - 1; i >= left; i--) {
                    cycle.add(grid[bottom][i]);
                }
            }

            // Traverse from bottom-1 to top+1 on left column, if left != right
            if (left < right) {
                for (int i = bottom - 1; i > top; i--) {
                    cycle.add(grid[i][left]);
                }
            }

            int n = cycle.size();
            if (n == 0) break;

            k = k % n;
            if (k == 0) {
                // No rotation needed, copy back
                int idx = 0;
                for (int i = left; i <= right; i++) {
                    result.get(top).add(cycle.get(idx++));
                }
                for (int i = top + 1; i <= bottom; i++) {
                    result.get(i).add(cycle.get(idx++));
                }
                if (top < bottom) {
                    for (int i = right - 1; i >= left; i--) {
                        result.get(bottom).add(cycle.get(idx++));
                    }
                }
                if (left < right) {
                    for (int i = bottom - 1; i > top; i--) {
                        result.get(i).add(cycle.get(idx++));
                    }
                }
                continue;
            }

            // Rotate the cycle
            List<Integer> rotated = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                rotated.add(cycle.get((i + k) % n));
            }

            int idx = 0;
            // Fill back into the grid
            for (int i = left; i <= right; i++) {
                result.get(top).add(rotated.get(idx++));
            }
            for (int i = top + 1; i <= bottom; i++) {
                result.get(i).add(rotated.get(idx++));
            }
            if (top < bottom) {
                for (int i = right - 1; i >= left; i--) {
                    result.get(bottom).add(rotated.get(idx++));
                }
            }
            if (left < right) {
                for (int i = bottom - 1; i > top; i--) {
                    result.get(i).add(rotated.get(idx++));
                }
            }

            top++;
            bottom--;
            left++;
            right--;
        }

        return result;
    }
}
}

File: MATTHEW/leetcode/AIcoding/1.Gemini-2.5/go_solutions/1914-cyclically-rotating-a-grid.go
package main

func rotateGrid(grid [][]int, k int) [][]int {
	m := len(grid)
	n := len(grid[0])

	for l := 0; l < m/2 && l < n/2; l++ {
		rStart := l
		rEnd := m - 1 - l
		cStart := l
		cEnd := n - 1 - l

		layerElements := []int{}

		// Extract elements of the current layer into a 1D slice
		// Top row (left to right)
		for c := cStart; c <= cEnd; c++ {
			layerElements = append(layerElements, grid[rStart][c])
		}
		// Right column (top to bottom, excluding top-right corner)
		for r := rStart + 1; r <= rEnd; r++ {
			layerElements = append(layerElements, grid[r][cEnd])
		}
		// Bottom row (right to left, excluding bottom-right corner)
		// This loop only runs if there is more than one row in the layer
		if rStart != rEnd {
			for c := cEnd - 1; c >= cStart; c-- {
				layerElements = append(layerElements, grid[rEnd][c])
			}
		}
		// Left column (bottom to top, excluding bottom-left and top-left corners)
		// This loop only runs if there is more than one column in the layer
		if cStart != cEnd {
			for r := rEnd - 1; r > rStart; r-- {
				layerElements = append(layerElements, grid[r][cStart])
			}
		}

		numElements := len(layerElements)
		if numElements == 0 {
			continue
		}

		effectiveK := k % numElements

		// Perform counter-clockwise rotation (left shift)
		rotatedLayerElements := append(layerElements[effectiveK:], layerElements[:effectiveK]...)

		// Place rotated elements back into the grid
		idx := 0

		// Top row (left to right)
		for c := cStart; c <= cEnd; c++ {
			grid[rStart][c] = rotatedLayerElements[idx]
			idx++
		}
		// Right column (top to bottom, excluding top-right corner)
		for r := rStart + 1; r <= rEnd; r++ {
			grid[r][cEnd] = rotatedLayerElements[idx]
			idx++
		}
		// Bottom row (right to left, excluding bottom-right corner)
		if rStart != rEnd {
			for c := cEnd - 1; c >= cStart; c-- {
				grid[rEnd][c] = rotatedLayerElements[idx]
				idx++
			}
		}
		// Left column (bottom to top, excluding bottom-left and top-left corners)
		if cStart != cEnd {
			for r := rEnd - 1; r > rStart; r-- {
				grid[r][cStart] = rotatedLayerElements[idx]
				idx++
			}
		}
	}

	return grid
}